### ISSUE TITLE
Improve code coverage for System.Xml.Xsl

### ISSUE THREAD

**Issue by sepidehMS** [dotnet/corefx#16650]
These are some of the bigger types with low coverage in `System.Xml.Xsl`. It would be great if we could add more tests to reach 80-90% coverage on these:

- [ ]  System.Xml.Xsl.Runtime.XmlCollation 45.6%
- [ ]  System.Xml.Xsl.Runtime.XmlQueryOutput 53.1%
- [ ]  System.Xml.Xsl.Runtime.XmlQueryRuntime 46.3%
- [ ]  System.Xml.Xsl.Runtime.XsltFunctions 10.3%
- [ ]  System.Xml.Xsl.Runtime.XsltLibrary 34.2%

Detailed coverage report:
https://ci.dot.net/job/dotnet_corefx/job/master/job/code_coverage_windows/Code_Coverage_Report/

**Comment by danmosemsft** [comment]
@hughbe I don't know whether this area is interesting to you, but there's a lot of XML code that could use better coverage. I bet there's plenty of relevant Mono tests that would port...

**Comment by danmosemsft** [comment]
Info on code coverage work:

1) Here's the all up report
https://ci.dot.net/job/dotnet_corefx/job/master/job/code_coverage_windows/Code_Coverage_Report/
Caveat - doesn't show coverage of any types whose implementation is in corelib (eg String, etc). If you want to look at that for now you must gather manually using instructions below.

2) Docs on code coverage work are here
https://github.com/dotnet/corefx/blob/master/Documentation/building/code-coverage.md
and some in
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md

3) There's innumerable places to add coverage -- we would love to focus on the most heavily used types and the least covered types to get the most benefit.

**Comment by WinCPP** [comment]
Hi @danmosemsft @sepidehMS Due to increasing workload at workplace, I won't be able to work fully on all of the areas mentioned for coverage in this issue. However, if ok, I would like to take one by one. May I work on "System.Xml.Xsl.Runtime.XsltFunctions 10.3%" to begin with. Do we need to create a separate issue for tracking purpose...? Thanks!

**Comment by danmosemsft** [comment]
@wincpp for sure! Nobody else is working on XML coverage so you can just go ahead.

**Comment by WinCPP** [comment]
@danmosemsft @sepidehMS I'm not able to figure out how to add a new Test project for code coverage. Went through [code-coverage.md](https://github.com/dotnet/corefx/blob/860b6f14c8f4186aecf061befa1fbe7b02349999/Documentation/building/code-coverage.md), but still at loss... My questions are very basic,
1. What should be folder hierarchy.
2. From docs it appears that one should not put project references. But then how are the types to be tested identified?
I want to work on coverage tests for System.Xml.Xsl.Runtime.XsltFunctions but I think I'm surely missing something.

The test projects in this area are quite _complex_ to understand the structure to use as template 😕

**Comment by hughbe** [comment]
> The test projects in this area are quite complex to understand the structure to use as template 😕

It's pretty nasty. I looked at this and couldn't really understand so gave up and moved on - good for you! The test code seems to be in src/System.Private.Xml/tests. There are a bunch of folders each with their own test library inside there. I can't really answer your other qs though

**Comment by WinCPP** [comment]
@sepidehMS ok. I think I got an answer why XsltFunctions has such a low coverage (10.3%). Using reflection, XsltLibrary.cs ([link](https://github.com/dotnet/corefx/blob/b0900bb7fd23d91e5d1f47539441b5d47b563d74/src/System.Private.Xml/src/System/Xml/Xsl/Runtime/XsltLibrary.cs#L30-L41)) gets hold of the functions in System.Xml.Xsl.Runtime.XsltFunctions. I suspect that would be the reason that although XsltLibrary has 100% coverage, XsltFunctions has very low coverage; whatever residual coverage might be due to direct access of few or more functions in some code path.

I think I will have to create an project that just tests the XsltFunctions methods? Tried doing that but I am not able to figure out how to add reference to System.Xml.Private.csproj in my test project as the doc that I mentioned earlier specifically mentions not to use project references. Thanks!

**Comment by danmosemsft** [comment]
@WinCPP I think all test projects automatically reference all the product. So you can just copy one of the existing csproj under \src\System.Private.Xml\tests.

**Comment by sepidehMS** [comment]
@WinCPP As mentioned by others, all Xml tests live under src\System.Private.Xml\tests and all test projects will automatically reference all product. (System.Xml.Xsl is exposed via System.Xml.ReaderWriter assembly) You can copy https://github.com/dotnet/corefx/blob/master/src/System.Private.Xml/tests/Xslt/XslTransformApi/System.Xml.Xsl.XslTransformApi.Tests.csproj and start from there.

**Comment by WinCPP** [comment]
@sepidehMS yup, I had followed all those steps and had used 

https://github.com/dotnet/corefx/blob/master/src/System.Private.Xml/tests/Xslt/XsltCompiler/XsltCompiler.Tests.csproj as the template. Even tried with the one that you have suggested but still I'm not able to access XsltFunctions in the System.Xml.Xsl.Runtime.XsltFunctions class in the tests. All that I can access is the System.Xml.Xsl namespace but nothing beyond that. Conversely, I checked if I can access System.Xml.Xsl.Runtime.XsltFunctions in the XsltTransformApi project, but it isn't accessible there either.

So I suspect, is XsltFunctions a publicly inaccessible class? In which case, how do I expose the same to the test project? I even tried adding reference to the System.Xml.Xsl project but the test project doesn't compile. Appreciate your advice.

**Comment by sepidehMS** [comment]
@WinCPP You are right. I totally missed it before, but all types in System.Xml.Xsl.Runtime are supposed to be called at runtime only and cannot be accessed directly from a user application, as these types are not exposed in the XML contracts (for the same reason). Given that, to increase code coverage for these types, you will need to write scenarios which make runtime calls into these APIs. I understand this might be more complicated, so if you pefer to start with something easier, we do need more coverage in other XML areas as well. You can take a look at: https://github.com/dotnet/corefx/issues/16680, https://github.com/dotnet/corefx/issues/16649, https://github.com/dotnet/corefx/issues/16646. Your help is appreciated :)
